addons/pos_stripe/models/pos_payment_method.py

```diff
@@ -0,0 +1,110 @@
+
+        for payment_method in self:
+            if not payment_method.stripe_serial_number:
+                continue
+            existing_payment_method = self.search([('id', '!=', payment_method.id),
+                                                   ('stripe_serial_number', '=', payment_method.stripe_serial_number)],
+                                                  limit=1)
+            if existing_payment_method:
+                raise ValidationError(_('Terminal %s is already used on payment method %s.')
+                                      % (payment_method.stripe_serial_number, existing_payment_method.display_name))
+
+    @api.model
+    def _get_stripe_secret_key(self):
+        stripe_secret_key = self.env['payment.acquirer'].search([('provider', '=', 'stripe')], limit=1).stripe_secret_key
+
+        if not stripe_secret_key:
+            raise ValidationError(_('Complete the Stripe onboarding.'))
+
+        return stripe_secret_key
+
+    @api.model
+    def stripe_connection_token(self):
+        endpoint = 'https://api.stripe.com/v1/terminal/connection_tokens'
+
+        try:
+            req = requests.post(endpoint, auth=(self._get_stripe_secret_key(), ''), timeout=TIMEOUT)
+        except ConnectionError:
+            raise UserError(_("Connection lost between Odoo and Stripe."))
+        except ConnectTimeout:
+            raise UserError(_("There are some connection issues between us and Stripe, try again later."))
+
+        if req.status_code == 401:
+            return {
+                'error': {
+                    'status_code': req.status_code,
+                    'message': req.text
+                }
+            }
+
+        return req.json()
+
+    @api.model
+    def stripe_payment_intent(self, currency, amount):
+        # For Terminal payments, the 'payment_method_types' parameter must include
+        # 'card_present' and the 'capture_method' must be set to 'manual'
+        endpoint = 'https://api.stripe.com/v1/payment_intents'
+
+        try:
+            data = ("currency=%s&amount=%s&payment_method_types[]=card_present&capture_method=manual") % (currency, int(amount * 100))
+            req = requests.post(endpoint, data=data, auth=(self._get_stripe_secret_key(), ''), timeout=TIMEOUT)
+        except ConnectionError:
+            raise UserError(_("Connection lost between Odoo and Stripe."))
+        except ConnectTimeout:
+            raise UserError(_("There are some connection issues between us and Stripe, try again later."))
+
+        if req.status_code == 401:
+            return {
+                'error': {
+                    'status_code': req.status_code,
+                    'message': req.text
+                }
+            }
+
+        return req.json()
+
+    @api.model
+    def stripe_capture_payment(self, paymentIntentId):
+        endpoint = ('https://api.stripe.com/v1/payment_intents/%s/capture') % (paymentIntentId)
```

Comment:
paymentIntentId should be url-encoded or you need to cast it as an int or something, you can't build a safe URL by concatenating arbitrary parameters

Reply:
🗒 checklist for safe `requests` calls:

- the endpoint URL is not injectable by callers, nor is the HTTP method, etc.
- any query/path parameters are properly escaped/encoded
- auth headers are safely handled
- a reasonable timeout is passed to each call
- the various error cases (timeout, 404, 500, etc.) are handled in some explicit manner
- the method does not leak low-level objects in the returned values, such as a Response or Session, only basic types